### PR TITLE
Combine standardisation and preprocessing defences

### LINKED ISSUE #625: Generalise framework-specific preprocessing to include estimator.preprocessing (standardisation)
The new framework-specific preprocessing defences of ART 1.4 for PyTorch and TensorFlow v2 allow the implementation of preprocessing defences and the backpropagation of gradients through them using the framework's methods and automated gradient calculation. 

The estimator of ART 1.4 support standardising the input with `(x - mean) / std` which is sufficient for some models for image classification but does not cover all standardisation methods of inputs.

This issue should build on the framework-specific preprocessing defences to provide support for general preprocessing (standardisation) functions.

These new methods for standardisation should become part of a new module `art.preprocessing`. This new module could also provide support for preprocessing discrete input data.

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request implements the new module `art.preprocessing` for general preprocessing tasks enabling framework-independent and framework-specific preprocessing of input data and the backpropagation of gradients through them using the framework's methods and automated gradient calculation.

As a first example `art.prerpocessing` includes the standardisation/normalisation of input `x` with `(x - mean) / std` framework-independent and framework-specific for TensorFlow v2 and PyTorch. This pull request replaces storing the current `preprocessing` as tuple of mean and standard deviation with an instance of `StandardisationMeanStd`, `StandardisationMeanStdPyTorch`, and `StandardisationMeanStdTensorFlowV2` of type `Preprocessor`.

This pull request also removes `_framework` methods in `PyTorchClassifier` and `TensorFlowV2Classifier` and integrates them into the respective standard methods, which now can be run with input of type `np.ndarray` or framework-specific tensors.

This pull request enables loss and class gradient backpropagation through model and preprocessing steps in one single operation instead of previously two separate steps for PyTorch and TensorFlow v2. That way the entire preprocessing including standardisation becomes differentiable in the same step as the model which should provide performance improvements. This pull request also enables framework-specific preprocessing (including standardisation) steps in the framework-specific Projected Gradient Descent (PGD) attacks in `ProjectedGradientDescentPyTorch` and `ProjectedGradientDescentTensorFlowV2`.

Fixes #625

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr_pytorch.py [hunk 1]
```diff
@@ -147,13 +147,7 @@ def __init__(
...
         from torch.autograd import Variable
 
-        if (
-            hasattr(estimator, "preprocessing")
-            and (estimator.preprocessing is not None and estimator.preprocessing != (0, 1))
-        ) or (
-            hasattr(estimator, "preprocessing_defences")
-            and (estimator.preprocessing_defences is not None and estimator.preprocessing_defences != [])
-        ):
+        if estimator.preprocessing:
```
Comment (minhitbk) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr_pytorch.py:150]: Are we working to enable the preprocessing of signals?

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py [hunk 2]
```diff
@@ -85,13 +85,7 @@ def __init__(
...
         :param verbose: Show progress bars.
         """
-        if (
-            hasattr(estimator, "preprocessing")
-            and (estimator.preprocessing is not None and estimator.preprocessing != (0, 1))
-        ) or (
-            hasattr(estimator, "preprocessing_defences")
-            and (estimator.preprocessing_defences is not None and estimator.preprocessing_defences != [])
-        ):
+        if not estimator.all_framework_preprocessing:
```
Comment (minhitbk) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:88]: Line 88-91: I think the if condition and the message in the raise are not related. Should we change the message to update the meaning of the condition?

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py [hunk 3]
```diff
@@ -84,13 +84,7 @@ def __init__(
...
         :param verbose: Show progress bars.
         """
-        if (
-            hasattr(estimator, "preprocessing")
-            and (estimator.preprocessing is not None and estimator.preprocessing != (0, 1))
-        ) or (
-            hasattr(estimator, "preprocessing_defences")
-            and (estimator.preprocessing_defences is not None and estimator.preprocessing_defences != [])
-        ):
+        if not estimator.all_framework_preprocessing:
```
Comment (minhitbk) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:87]: Line 87-90: I think the if condition and the message in the raise are not related. Should we change the message to update the meaning of the condition?

File: art/defences/preprocessor/preprocessor.py [hunk 4]
```diff
@@ -162,6 +164,55 @@ def estimate_forward(self, x: "torch.Tensor", y: Optional["torch.Tensor"] = None
         """
         raise NotImplementedError
 
+    def __call__(self, x: np.ndarray, y: Optional[np.ndarray] = None) -> Tuple[np.ndarray, Optional[np.ndarray]]:
+        """
+        Apply local spatial smoothing to sample `x`.
```
Comment (minhitbk) [inline art/defences/preprocessor/preprocessor.py:169]: Line 169-173: I think the documents are not updated. It is for label smoothing.

File: art/defences/preprocessor/preprocessor.py [hunk 5]
```diff
@@ -193,3 +244,54 @@ def estimate_forward(self, x: "tf.Tensor", y: Optional["tf.Tensor"] = None) -> "
         :return: Preprocessed data.
         """
         raise NotImplementedError
+
+    def __call__(self, x: np.ndarray, y: Optional[np.ndarray] = None) -> Tuple[np.ndarray, Optional[np.ndarray]]:
+        """
+        Apply local spatial smoothing to sample `x`.
```
Comment (minhitbk) [inline art/defences/preprocessor/preprocessor.py:250]: Line 250-254: I think the documents are not updated. It is for label smoothing.

File: art/estimators/classification/pytorch.py [hunk 6]
```diff
@@ -510,17 +511,30 @@ def loss_gradient(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
         import torch  # lgtm [py/repeated-import]
 
         # Apply preprocessing
-        x_preprocessed, y_preprocessed = self._apply_preprocessing(x, y, fit=False)
+        if self.all_framework_preprocessing:
+            if isinstance(x, torch.Tensor):
```
Comment (minhitbk) [inline art/estimators/classification/pytorch.py:515]: Line 515-520: I think the functions `torch.tensor` and `torch.fromnumpy` have similar functionality, so we don't need an `if..else` condition here.

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    def apply_fit(self) -> bool:
+        return self._apply_fit
+
+    @property
+    def apply_predict(self) -> bool:
+        return self._apply_predict
+
+    def forward(self, x: "tf.Tensor", y: Optional["tf.Tensor"] = None) -> Tuple["tf.Tensor", Optional["tf.Tensor"]]:
+        """
+        Apply local spatial smoothing to sample `x`.
```
Comment (minhitbk) [inline art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py:69]: I think the document here should be updated.

File: tests/estimators/classification/test_input_filter.py [hunk 8]
```diff
@@ -57,304 +57,304 @@ def test_fit(self):
...
-                0.01474165,
-                0.0,
-                0.0,
-                0.0,
-                0.0,
-                0.0,
-            ]
-        )
-        np.testing.assert_array_almost_equal(gradients[0, :, 14, 0], expected_gradients_2, decimal=4)
+    # def test_class_gradient(self):
```
Comment (minhitbk) [inline tests/estimators/classification/test_input_filter.py:60]: Why do you comment out the test, should we remove it if it is not necessary?

**Review by minhitbk — CHANGES_REQUESTED** [review]
Hi @beat-buesser , I just have a few minor comments on this PR. Please take a look.

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr_pytorch.py (hunk 1 above)
Comment (beat-buesser) [reply art/attacks/evasion/imperceptible_asr/imperceptible_asr_pytorch.py:150]: The former `preprocessing_defences` is being integrated into `preprocessing` with the goal to have only one preprocessing abstraction and not distinguish between preprocessing and preprocessing defences anymore. Based on offline discussion with @minhitbk we don't need this check anymore and I'll remove it with a new commit.

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py (hunk 2 above)
Comment (beat-buesser) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_pytorch.py:88]: I agree, fixed with 53baf291bd0fa5749bffeac752b4d77a1ee5ceea

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py (hunk 3 above)
Comment (beat-buesser) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_tensorflow_v2.py:87]: I agree, fixed with 53baf291bd0fa5749bffeac752b4d77a1ee5ceea

File: art/defences/preprocessor/preprocessor.py (hunk 4 above)
Comment (beat-buesser) [reply art/defences/preprocessor/preprocessor.py:169]: I agree, fixed with 1d7b82217804353137a7e9c916459f8641c80632

File: art/defences/preprocessor/preprocessor.py (hunk 5 above)
Comment (beat-buesser) [reply art/defences/preprocessor/preprocessor.py:250]: I agree, fixed with 1d7b82217804353137a7e9c916459f8641c80632

File: art/estimators/classification/pytorch.py (hunk 6 above)
Comment (beat-buesser) [reply art/estimators/classification/pytorch.py:515]: Nice, I was not aware of this, fixed with 3fc9975df395814b6cb9f2d7bedecbbf336a44e0

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py (hunk 7 above)
Comment (beat-buesser) [reply art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py:69]: I agree, fixed with 4e40f50a2f5fe3698be8d907a9c5fd71619e7998

File: tests/estimators/classification/test_input_filter.py (hunk 8 above)
Comment (beat-buesser) [reply tests/estimators/classification/test_input_filter.py:60]: I would like to keep it because we plan to reintroduce the feature later.

**Review by minhitbk — APPROVED** [review]
